### PR TITLE
Add configuration to delay the first start of Push until Push.setEnabled(true) is called from JS

### PR DESCRIPTION
**Pull request by guperrot**
Things to consider before you submit the PR:

* [x] Has `CHANGELOG.md` been updated?
* [x] Are tests passing locally?
* [x] Are the files formatted correctly?
* [x] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?

## Description

To delay start push with this change:

* Add EnablePushInJavascript YES to AppCenter-Config.plist
* Add `"enable_push_in_javascript": true` to appcenter-config.json
* Call `Push.setEnabled(true)` from Javascript when developer wants to start Push.

Since this is only for the first time we want to delay Push, the code also remembers the first call to `Push.setEnabled(true)` and will start Push automatically the next time.

Documentation PR: https://github.com/MicrosoftDocs/appcenter-docs-pr/pull/1000

Requested in #287

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,11 +1,15 @@
...
-## Version 1.10.1 (under active development)
+## Version 1.11.0 (under active development)
 
 ### AppCenter
 
 - Introduce new `LogLevel` constants, deprecating old ones.
 
+### AppCenterPush
+
+- **[Feature]** Add manual configuration to delay the first start of Push module.
```
Comment (dhei) [inline CHANGELOG.md]: I suggest something like "Support delaying Push notification permission dialog". The original feature request is at https://github.com/Microsoft/AppCenter-SDK-React-Native/issues/287.

File: appcenter-analytics/android/build.gradle [hunk 2]
```diff
@@ -21,6 +21,6 @@ dependencies {
     compile 'com.facebook.react:react-native:+'
     compile 'com.microsoft.appcenter:appcenter-analytics:1.10.0'
 
-    //compile project(':AppCenterReactNativeShared')   // For testing with TestApp
-    compile 'com.microsoft.appcenter.reactnative:appcenter-react-native:1.10.0'
+    compile project(':AppCenterReactNativeShared')   // For testing with TestApp
```
Comment (dhei) [inline appcenter-analytics/android/build.gradle]: revert this change.

File: appcenter-push/android/src/main/java/com/microsoft/appcenter/reactnative/push/AppCenterReactNativePushModule.java [hunk 3]
```diff
@@ -10,18 +10,39 @@
...
         AppCenterReactNativeShared.configureAppCenter(application);
         if (AppCenter.isConfigured()) {
-            AppCenter.start(Push.class);
+            boolean startPush = true;
+            boolean enablePushInJavascript = AppCenterReactNativeShared.getConfiguration().optBoolean(ENABLE_PUSH_IN_JAVASCRIPT);
+            if (enablePushInJavascript) {
+
+                /*
+                 * TODO expose a way to post command in background looper in native SDK to avoid accessing storage directly here.
+                 * Storage might not be ready yet.
```
Comment (jwallra) [inline appcenter-push/android/src/main/java/com/microsoft/appcenter/reactnative/push/AppCenterReactNativePushModule.java]: What happens if storage is not ready?  Would we crash, or would the push configuration just not be set?

File: appcenter-push/android/src/main/java/com/microsoft/appcenter/reactnative/push/AppCenterReactNativePushModule.java [hunk 4]
```diff
@@ -37,11 +58,20 @@ public String getName() {
...
+            AppCenter.start(Push.class);
+            mPushStarted = true;
+        }
         Push.setEnabled(enabled).thenAccept(new AppCenterConsumer<Void>() {
 
             @Override
             public void accept(Void result) {
+                if (enabled) {
+
+                    /* TODO expose a way to post command in background looper in native SDK to avoid accessing storage directly here. */
```
Comment (jwallra) [inline appcenter-push/android/src/main/java/com/microsoft/appcenter/reactnative/push/AppCenterReactNativePushModule.java]: Does this have the same condition ("storage might not be ready yet") as above?  If so, let's include that in the comment.

File: appcenter-analytics/android/build.gradle (hunk 2 above)
Comment (guperrot) [reply appcenter-analytics/android/build.gradle]: Won't be able to compile if I do that, we need to keep that until release.

File: appcenter-push/android/src/main/java/com/microsoft/appcenter/reactnative/push/AppCenterReactNativePushModule.java (hunk 3 above)
Comment (guperrot) [reply appcenter-push/android/src/main/java/com/microsoft/appcenter/reactnative/push/AppCenterReactNativePushModule.java]: We make the storage ready in the next line.

File: appcenter-push/android/src/main/java/com/microsoft/appcenter/reactnative/push/AppCenterReactNativePushModule.java (hunk 4 above)
Comment (guperrot) [reply appcenter-push/android/src/main/java/com/microsoft/appcenter/reactnative/push/AppCenterReactNativePushModule.java]: @jwallra no as storage is initialized in the constructor of this class.
@achocron it should not be a blocker for this work (we already read a file in shared directly = disk access so would require more changes to the RN SDK overall to comply with strict mode) item but yes we can add a work item for that.

File: appcenter-analytics/android/build.gradle (hunk 2 above)
Comment (guperrot) [reply appcenter-analytics/android/build.gradle]: Your feedback is addressed in https://github.com/Microsoft/AppCenter-SDK-React-Native/pull/480/commits/a3adae58e3a62b77aa2185e7ff974beb9a24c3db

File: scripts/bump-android-sdk-version.sh [hunk 5]
```diff
@@ -35,19 +35,14 @@ echo "React-Native Android version $oldAndroidSdkVersion will be updated to $new
...
-
-gradleFileContent="$(cat ./appcenter-analytics/android/build.gradle)"
-echo "${gradleFileContent/com.microsoft.appcenter\:appcenter-analytics\:$oldAndroidSdkVersion/com.microsoft.appcenter:appcenter-analytics:$newAndroidSdkVersion}" > ./appcenter-analytics/android/build.gradle
-
-gradleFileContent="$(cat ./appcenter-push/android/build.gradle)"
-echo "${gradleFileContent/com.microsoft.appcenter\:appcenter-push\:$oldAndroidSdkVersion/com.microsoft.appcenter:appcenter-push:$newAndroidSdkVersion}" > ./appcenter-push/android/build.gradle
-
-gradleFileContent="$(cat ./AppCenterReactNativeShared/android/build.gradle)"
-echo "${gradleFileContent/com.microsoft.appcenter\:appcenter\:$oldAndroidSdkVersion/com.microsoft.appcenter:appcenter:$newAndroidSdkVersion}" > ./AppCenterReactNativeShared/android/build.gradle
+for file in \
```
Comment (dhei) [inline scripts/bump-android-sdk-version.sh]: I like this improvement in bump-android-sdk-version.sh and bump-wrapper-sdk-version.sh. We should do the same for bump-ios-sdk-version.sh.

**Comment by jdnichollsc** [comment]
`EnablePushInJavascript` is not working for iOS, the event **onPushNotificationReceived** is not executed in that platform

**Comment by guperrot** [comment]
@jdnichollsc it works for us in TestApp from this repo but it looks like your issue is caused by something else since you commented on #505.

**Comment by jdnichollsc** [comment]
@guperrot thanks, I'm reviewing the code of AppCenter but I can't see anything

**Comment by guperrot** [comment]
We will investigate and track in #505 even if not related (can still rename the title there when we find the cause).

Also you mentioned in #505 that version 1.8.0 of the SDK also has the issue so that means it's not caused by this new configuration feature and is a older issue.
